### PR TITLE
feat(query): migrate regexp func to func-v2

### PR DESCRIPTION
**Pull request by b41sh**
I hereby agree to the terms of the CLA available at: https://databend.rs/dev/policies/cla/

## Summary

migrate regexp functions to function-v2
```
regexp_instr
regexp_like
regexp_replace
regexp_substr
```

Part of #6766

### CODE REVIEW

**Comment by BohuTANG** [comment]
Conflicting files
src/query/functions-v2/src/scalars/string.rs

File: src/query/codegen/src/writes/register.rs [hunk 1]
```diff
@@ -12,7 +12,7 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-const MAX_ARGS: usize = 5;
+const MAX_ARGS: usize = 6;
```
Comment (sundy-li) [inline src/query/codegen/src/writes/register.rs:15]: 6 leads codebloat ...

File: src/query/functions-v2/src/scalars/string.rs [hunk 2]
```diff
@@ -746,6 +751,393 @@ pub fn register(registry: &mut FunctionRegistry) {
...
+        vectorize_with_builder_6_arg::<
+            StringType,
+            StringType,
+            NumberType<i64>,
+            NumberType<i64>,
+            NumberType<i64>,
+            StringType,
+            NumberType<u64>
+        >(|source, pat, pos, occur, ro, mt, output| {
+            let instr = regexp::regexp_instr_scalar(source, pat, Some(pos), Some(occur), Some(ro), Some(mt))?;
```
Comment (sundy-li) [inline src/query/functions-v2/src/scalars/string.rs:779]: can we use `index_unchecked` to replace `vectorize_with_builder_6_arg` ?

File: src/query/functions-v2/tests/it/scalars/testdata/string.txt [hunk 3]
```diff
@@ -3138,3 +3138,773 @@ evaluation (internal):
...
+| Domain | {""..="dog cat dog"}      | {""..="dog"} | Unknown |
+| Row 0  | "dog cat dog"             | "dog"        | 1       |
+| Row 1  | "aa aaa aaaa aa aaa aaaa" | "a{2}"       | 1       |
+| Row 2  | ""                        | ""           | 0       |
++--------+---------------------------+--------------+---------+
+evaluation (internal):
++--------+-------------------------------------------------------------------------------------------------------------------------+
+| Column | Data                                                                                                                    |
++--------+-------------------------------------------------------------------------------------------------------------------------+
+| source | StringColumn { data: 0x646f672063617420646f676161206161612061616161206161206161612061616161, offsets: [0, 11, 34, 34] } |
```
Comment (sundy-li) [inline src/query/functions-v2/tests/it/scalars/testdata/string.txt:3202]: data is formated unreadable. cc @andylokandy

File: src/query/functions-v2/tests/it/scalars/testdata/string.txt [hunk 4]
```diff
@@ -3138,3 +3138,957 @@ evaluation (internal):
 +--------+--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------+
 
 
+ast            : regexp_instr('dog cat dog', 'dog', 1)
```
Comment (sundy-li) [inline src/query/functions-v2/tests/it/scalars/testdata/string.txt:3141]: can we move `regexp_xxx` functions to a new test file?

Comment (b41sh) [reply src/query/functions-v2/tests/it/scalars/testdata/string.txt:3141]: ok

**Review by sundy-li — APPROVED** [review]

File: src/query/functions-v2/src/scalars/string_multi_args.rs [hunk 5]
```diff
@@ -279,6 +286,330 @@ pub fn register(registry: &mut FunctionRegistry) {
...
+        } else if args_type.len() == 5 {
+            vec![
+                DataType::String,
+                DataType::String,
+                DataType::Number(NumberDataType::Int64),
+                DataType::Number(NumberDataType::Int64),
+                DataType::Number(NumberDataType::Int64),
+            ]
+        } else {
+            vec![
```
Comment (andylokandy) [inline src/query/functions-v2/src/scalars/string_multi_args.rs:319]: How about using match?

File: src/query/functions-v2/src/scalars/string_multi_args.rs [hunk 6]
```diff
@@ -348,3 +679,515 @@ fn char_fn(args: &[ValueRef<AnyType>], _: &GenericMap) -> Result<Value<AnyType>,
...
+    };
+
+    let size = len.unwrap_or(1);
+    let mut key: Vec<Vec<u8>> = Vec::new();
+    let mut map: HashMap<Vec<Vec<u8>>, Regex> = HashMap::new();
+    let mut builder = NumberColumnBuilder::with_capacity(&NumberDataType::UInt64, size);
+    for idx in 0..size {
+        unsafe {
+            let source = source_arg.index_unchecked(idx);
+            let pat = pat_arg.index_unchecked(idx);
```
Comment (andylokandy) [inline src/query/functions-v2/src/scalars/string_multi_args.rs:719]: Are we going to cache the regex for scalar pat later?

Comment (b41sh) [reply src/query/functions-v2/src/scalars/string_multi_args.rs:679]: `regex` is cached in a map, if `pat` and `match type` is equal.

**Review by andylokandy — APPROVED** [review]

**Pull request merged**